// nodeFlows/workflows/Inactive/pizzaOrderingWorkflow - withstockcheck.js
//Workflow to simulate the taking of a pizza order

var workflowDefinition = {
    Name: "Workflow1",
    Steps: [
        { Name: "Step0",
            Type: "Init",
            args:
                { status: "Ordered", currentStepName: "OrderReceived", deliveryType: "Delivery" }
        },
        { Name: "OrderReceived",
            Type: "writeConsoleActivity",
            args:
                { text: "Workflow has been started" },
            nextStep: "EnterIntoComputer"
        },
        { Name: "EnterIntoComputer",
            Type: "writeConsoleActivity",
            args:
                { text: "Workflow has been started2" },
            nextStep: "SendToCook"
        },
        { Name: "SendToCook",
            Type: "createChildWorkflow",
            args:
                { workflowDefinition: "PizzaOrdering/cookWorkflow" },
            nextStep: "WaitForCooking"
        },
        { Name: "WaitForCooking",
            Type: "waitForActivity",
            args:
                { Condition: "workflow.status=='Cooked'" },
            nextStep: "CookingComplete"
        },
        { Name: "CookingComplete",
            Type: "ifElseActivity",
            args:
                { text: "ifElse activity Step",
                    branches: [
                    { Condition: "workflow.deliveryType=='Delivery'", Step: "SendForDelivery" },
                    { Condition: "workflow.deliveryType=='Takeout'", Step: "WaitForCustomer" },
                    { Condition: null, Step: "WaitForCustomer"}  //Default
                    ]
                },
            nextStep: null
        },
        { Name: "SendForDelivery",
            Type: "createChildWorkflow",
            args:
                { workflowDefinition: "PizzaOrdering/deliveryWorkflow" },
            nextStep: "WaitForDelivery"
        },
        { Name: "WaitForDelivery",
            Type: "waitForActivity",
            args:
                { Condition: "workflow.status=='Delivered'" },
            nextStep: "CheckStocks"
        },
        { Name: "WaitForCustomer",
            Type: "writeConsoleActivity",
            args:
                { text: "Waiting for Customer Pickup" },
            nextStep: "CheckStocks"
        },
        { Name: "CheckStocks",
            Type: "callUrlActivity",
            args:
                { 
                host: "query.yahooapis.com",
                path: "/v1/public/yql?q=select%20*%20from%20yahoo.finance.quotes%20where%20symbol%20in%20(%22YHOO%22%2C%22AAPL%22%2C%22GOOG%22%2C%22MSFT%22)&format=json&env=store%3A%2F%2Fdatatables.org%2Falltableswithkeys&callback=cbfunc" 
                },
            nextStep: "Complete"
        },
        { Name: "Complete",
            Type: "completeWorkflow",
            args:
                {},
            nextStep: null
        }
    ]
}


exports.workflowDefinition = workflowDefinition;
