// nodeFlows/workflows/Tests/testRepeatingPizzaOrders1000.js


//Repeat the same pizza order 1000 times

var workflowDefinition = {
    Name: "testLoader",
    Steps: [
        { Name: "Step0",
            Type: "Init",
            args:
                { currentStepName: "LogActivity", status: "Ordered" }
        },
        { Name: "LogActivity",
            Type: "writeLogActivity",
            args:
                { text: "Getting a pizza order..." }
        },
        { Name: "StartOrder",
            Type: "createChildWorkflow",
            args:
                { workflowDefinition: "PizzaOrdering/pizzaOrderingWorkflow", parameters: {deliveryType: "Delivery"} }
        },
        { Name: "WaitForTimeout",
            Type: "waitForTimeout",
            args:
                { timeout: 40 }
        },
        { Name: "IncrementIterations",
            Type: "incrementWorkingStorage",
            active: true,
            args:
                { valueName: "OrdersPlaced", incrementAmount:1 }
        },
        { Name: "HowManyIterations",
            Type: "getWorkingStorageValue",
            active: true,
            args:
                { valueName: "OrdersPlaced", workflowProperty: "OrdersPlaced" }
        },
        { Name: "CompleteIf1000Iterations",
            Type: "ifElseActivity",
            args:
                { text: "ifElse activity Step",
                    branches: [
                    { Condition: "this.OrdersPlaced>1000", Step: "Complete" },
                    { Condition: null, Step: "Restarting"}  //Default
                    ]
                },
            nextStep: null
        },
        { Name: "Restarting",
            Type: "writeLogActivity",
            args:
                { text: "Getting a pizza order..." },
            nextStep: "LogActivity"
        },
        { Name: "Complete",
            Type: "completeWorkflow",
            args:
                {},
            nextStep: null
        }

    ]
}

exports.workflowDefinition = workflowDefinition;
