// nodeFlows/workflows/Tests/testURLCall.js




var workflowDefinition = {
    Name: "testUrlCall",
    Steps: [
        { Name: "Step0",
            Type: "Init",
            args:
                { currentStepName: "LogActivity", status: "Ordered" },
            nextStep: "LogActivity"
        },
        { Name: "LogActivity",
            Type: "writeConsoleActivity",
            args:
                { text: "testing URL call has started" },
            nextStep: "StartOrder"
        },
        { Name: "StartOrder",
            Type: "postUrl",
            args:
                { 
                host: "query.yahooapis.com",
                path: "/v1/public/yql?q=select%20*%20from%20yahoo.finance.quotes%20where%20symbol%20in%20(%22YHOO%22%2C%22AAPL%22%2C%22GOOG%22%2C%22MSFT%22)&format=json&env=store%3A%2F%2Fdatatables.org%2Falltableswithkeys&callback=cbfunc" 
                },
            nextStep: "WaitABit"
        },
        { Name: "WaitABit",
            Type: "waitForTimer",
            args:
                { waitTime: 10000 },
            nextStep: "GotoStep"
        },
        { Name: "GotoStep",
            Type: "gotoStep",
            args:
                { step: "LogActivity" },
            nextStep: null
        },
        { Name: "Complete",
            Type: "completeWorkflow",
            args:
                {},
            nextStep: null
        }

    ]
}

exports.workflowDefinition = workflowDefinition;
