// nodeFlows/workflows/PizzaOrdering/cookWorkflow.js
//Workflow to simulate the cooks activities

var workflowDefinition = {
    Name: "PizzaCooker",
    workers: 6,
    Steps: [
        { Name: "StartWorkflow",
            Type: "Init",
            args:
                { status: "Received", currentStepName: "PizzaOrderReceived" }
        },
        { Name: "PizzaOrderReceived",
            Type: "writeLogActivity",
            args:
                { text: "Cook Workflow has been started" }
        },
        { Name: "PreparePizza2",
            Type: "writeLogActivity",
            args:
                { text: "Preparing the pizza" }
        },
        { Name: "GotoStep",
            Type: "gotoStep",
            active: true,
            args:
                { step: "AddPepperonis" },
            nextStep: null
        },
        { Name: "AddPepperonis",
            Type: "writeLogActivity",
            args:
                { text: "Adding Pepperonis" }
        },
        { Name: "HowManyPizzasAreCooking",
            Type: "getWorkingStorageValue",
            active: true,
            args:
                { valueName: "PizzasInOven", workflowProperty: "PizzasInOven" }
        },
        { Name: "SmokeBreak",
            Type: "writeLogActivity",
            active: false,
            args:
                { text: "The cook is taking a smoke break" }
        },
        { Name: "PutPizzaInOven",
            Type: "incrementWorkingStorage",
            active: true,
            args:
                { valueName: "PizzasInOven", incrementAmount:1 }
        },
        { Name: "CookPizza",
            Type: "waitForTimer",
            persistAfter: true,
            args:
                { waitTime: 1000 }
        },
        { Name: "PizzaReady",
            Type: "updateInitiatingWorkflowActivity",
            args:
                { 'status': 'Cooked' },
            success:
                {statistic: 'PizzasCooked'}
        },
        { Name: "HowManyPizzasAreCooking2",
            Type: "getWorkingStorageValue",
            active: true,
            args:
                { valueName: "PizzasInOven", workflowProperty: "PizzasInOven" }
        },
        { Name: "RemovePizzaFromOven",
            Type: "incrementWorkingStorage",
            active: true,
            args:
                { valueName: "PizzasInOven", incrementAmount:-1 }
        },
        { Name: "Complete",
            Type: "completeWorkflow",
            args:
                {},
            nextStep: null
        }

    ]
}


exports.workflowDefinition = workflowDefinition;
