// nodeFlows/activities/postUrl.js
var nodeFlows = require('../nodeFlows');
var http = require("http");

exports.executeActivity = executeActivity;

function executeActivity(workflow, args, callback) {

/*
  var post_data = JSON.stringify({ 
      workflowDefinition : 'PizzaOrdering/pizzaOrderingWorkflow', 
      args: '' 
  }); 
*/
  var post_data = "workflowDefinition=PizzaOrdering/pizzaOrderingWorkflow&args=''" 
 
  // An object of options to indicate where to post to 
  var post_options = { 
      host: 'localhost', 
      port: 13465, 
      path: '/rest/workflows/create', 
      method: 'POST', 
      headers: { 
          'Content-Type': 'application/x-www-form-urlencoded', 
          'Content-Length': post_data.length,
          'user-agent': 'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)'
      } 
  }; 

  // Set up the request 
  var post_req = http.request(post_options, function(res) { 
      res.setEncoding('utf8'); 
      res.on('data', function (chunk) { 
          nodeFlows.log('Response: ' + chunk); 
          callback({ success: true, activity: "postUrl" });
      }); 
  }); 

  // post the data 
  post_req.write(post_data); 
  post_req.end(); 

/*
    var x = "";
    http.get({ host: args.host, path: args.path }, function(res) {
        res.on('data', function(d) {
            x += d.toString();
            //console.log(d.toString() + "<br>");
            //console.log("NoString:" + d.toString() + "<br>");
            //console.log(d.query);
        })
        res.on('end', function() {
            var l = JSON.parse(x.substring(7, x.length - 2));
            console.log("Results:" + x);
            console.log("Result:" + l.query.results.quote[0].symbol + '');
            callback({ success: true });
        });

    });
*/

}
